### PR TITLE
controller: Migration listener

### LINKED ISSUE #2235: controller: "cached plan must not change result type"
```text
ERROR: cached plan must not change result type (SQLSTATE 0A000)
```

Sounds like a race between the DB migrations and preparing statements when creating connections.

https://gist.github.com/anonymous/2eedd040b6bf05bcd1da

### PR DESCRIPTION
**Pull request by josephglanville**
This change implements a listen/notify setup on the `schema_migrations` table in order to reset the connection pool after a migration takes place. This is to ensure any session level cached or prepared query plans can be refreshed.

Refs #2295 
Closes #2235

### CODE REVIEW

**Comment by josephglanville** [comment]
A similar change will be needed for the router if/when it's changed to use prepared queries, this is why I implemented the core logic in `pkg/postgres` instead of in the controller itself.

Right now the datastore actually generates the queries semi on the fly which is probably terrible for performance with pgx. (which anonymously prepares queries on exec if they aren't prepared already..)
That said because it doesn't use prepared queries means it's probably immune from these sorts of errors right now.

File: controller/controller.go [hunk 1]
```diff
@@ -83,6 +83,11 @@ func main() {
 		}
 	}()
 
+	// Listen for database migration, reset connpool on new migration
+	go func() {
+		postgres.ResetOnMigration(db, logger, doneCh)
```
Comment (titanous) [inline controller/controller.go]: No need to wrap in an anonymous function, you can just do `go postgres.ResetOnMigration...`

File: pkg/postgres/migrate.go [hunk 2]
```diff
@@ -65,3 +73,30 @@ func (m Migrations) Migrate(db *DB) error {
...
+				}
+				log.Warn("new schema migration, resetting conn pool", "id", n.Payload)
+				db.Reset()
+			case _ = <-doneCh:
+				listener.Close()
+				return
+			}
+		}
+	}
+	return
```
Comment (titanous) [inline pkg/postgres/migrate.go]: I think this return is uncessary.

File: pkg/postgres/migrate.go [hunk 3]
```diff
@@ -65,3 +73,30 @@ func (m Migrations) Migrate(db *DB) error {
...
+			case n, ok := <-listener.Notify:
+				if !ok {
+					log.Warn("migration listener disconnected, reconnecting in 5 seconds")
+					time.Sleep(5 * time.Second)
+					listener.Close()
+					break outer
+				}
+				log.Warn("new schema migration, resetting conn pool", "id", n.Payload)
+				db.Reset()
+			case _ = <-doneCh:
```
Comment (titanous) [inline pkg/postgres/migrate.go]: No need for `_ =`

**Comment by josephglanville** [comment]
Comments addressed, cheers for review, still getting used to programming with channels.

File: pkg/postgres/migrate.go (hunk 2 above)
Comment (titanous) [inline pkg/postgres/migrate.go]: Was the `return` required?

**Comment by titanous** [comment]
LGTM

**Pull request merged**